### PR TITLE
Add travis support for basic start of spyder

### PR DESCRIPTION
**Pull request by goanpeca**
Related to #2268

Todo:
- [x] Use bootstrap options
- [x] Change user help message
- [x] Include optional dependencies

### CODE REVIEW

**Comment by goanpeca** [comment]
Just a basic start, so are I got this versions of python running

https://travis-ci.org/goanpeca/spyder/builds/56129095

@blink1073, @ccordoba12

**Comment by blink1073** [comment]
This is awesome @goanpeca!  @ccordoba12, since you are the owner of the repo you'll have to be the one to enable it on Travis. 

- log into travis-ci.org
- sign in with Github
- click on your avatar on the top right and select Accounts
- slide the spyder repo on

If you don't see the spyder repos, you'll have to go to the spyder-ide org on github and set yourself to "public" on the owner team.

**Comment by goanpeca** [comment]
I think this PR is enough on its own, for the actual tests I prefer to open a different PR, what do you guys think?

File: continuous_integration/travis_install.sh [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    conda config --add channels dsdale24;
+    conda create -q -n test-environment python=$PY_VERSION sphinx sip qt5 pyqt5;
+    ldd "$HOME/miniconda/envs/test-environment/lib/qt5/plugins/platforms/libqxcb.so"
+
+    # libxcb-atom1-dev libxcb-event1-dev libxcb-icccm1-dev
+    # ldd /home/goanpeca/anaconda/envs/test-environment/lib/qt5/plugins/platforms/libqxcb.so
+    #	
+
+  elif [ "$USE_QT_API" = "PyQt4" ]; then
+    conda create -q -n test-environment python=$PY_VERSION sphinx sip qt pyqt;
```
Comment (ccordoba12) [inline continuous_integration/travis_install.sh]: We should include here more packages. At least `ipython-qtconsole`, `matplotlib` and `pandas`.

File: continuous_integration/travis_install.sh [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    for LIB in ${LIBS[@]}
+    do
+        sudo ln -sf $LIB_SYSTEM_PATH/$LIB $LIB_VIRTUALENV_PATH/$LIB
+    done
+}
+
+install_qt5()
+{
+  echo "not yet!"
+  sudo add-apt-repository --yes ppa:ubuntu-sdk-team/ppa
```
Comment (ccordoba12) [inline continuous_integration/travis_install.sh]: Why is this necessary to use that ppa? Qt5 is already part of the official packages

Comment (goanpeca) [inline continuous_integration/travis_install.sh]: Not in ubuntu 12.04 (travis), besides... as it says.. not yet supported :)!

File: continuous_integration/travis_install.sh (hunk 1 above)
Comment (goanpeca) [inline continuous_integration/travis_install.sh]: Sure, any other packages besides the ones you mention?

File: continuous_integration/travis_install.sh (hunk 2 above)
Comment (ccordoba12) [inline continuous_integration/travis_install.sh]: Ok, I forgot that Travis uses 12.04 ;-)

File: spyderlib/cli_options.py [hunk 3]
```diff
@@ -39,5 +40,8 @@ def get_options():
     parser.add_option('--profile', action='store_true', default=False,
                       help="Profile mode (internal test, "
                            "not related with Python profiling)")
+    parser.add_option('--travis', dest="travis", action='store_true',
+                      default=False,
+                      help="Run a Qtimer to stop the GUI app in travis-ci")
```
Comment (ccordoba12) [inline spyderlib/cli_options.py]: Mmmh, I think this option's name and help should be rephrased, given that it's going to show when users do `spyder --help`.

What about

    '--test', ....
    help="Closes the application after 30 secs. Useful only for testing purposes".

Comment (goanpeca) [inline spyderlib/cli_options.py]: Yes it makes sense, I will change

File: continuous_integration/travis_install.sh (hunk 1 above)
Comment (ccordoba12) [inline continuous_integration/travis_install.sh]: No, not the minimal versions, those are usually too old. We need to test things against their latest versions.

I'd say we should test against all packages mentioned in *Optional dependencies*,

**Comment by ccordoba12** [comment]
Thanks a lot for working on this @goanpeca, besides the couple of comments I made, I think this is ready for merging :-)

As you said, we can leave moving our tests to a `tests` directory for another PR.

File: spyderlib/cli_options.py (hunk 3 above)
Comment (goanpeca) [inline spyderlib/cli_options.py]: hmmm maybe `test-travis`? is better? or

`test-ci` `Closes the application after 30 secs. Useful only for continuous integration (travis-ci)  testing purposes"

Comment (ccordoba12) [inline spyderlib/cli_options.py]: What if instead of adding this option to the main interface, we add it to `boostrap.py`?

I think it makes more sense there, don't you think?

Comment (ccordoba12) [inline spyderlib/cli_options.py]: The thing is, `travis` or `ci` are technical terms our users don't need to know or understand to the read `spyder --help`.

So if you prefer the option to remain as part of our main cli options, I think using `--test` for the option (and the help I suggested before) is just fine.

Comment (goanpeca) [inline spyderlib/cli_options.py]: I agree on making it a bootstrap option instead but how do I pass it to the main spyder app (where I need to set the timer?)

Comment (ccordoba12) [inline spyderlib/cli_options.py]: By using an environment variable. Take a look at how I define and use the `DEV` variable in `bootstrap.py` and then in `spyder.py`

**Comment by goanpeca** [comment]
Ok so I made the changes suggested,

@ccordoba12, can you go to https://travis-ci.org/spyder-ide/spyder/settings and activate testing only if a .yml is present?

**Comment by goanpeca** [comment]
@blink1073 , can you take a look at the build in my test repo?

https://travis-ci.org/goanpeca/spyder/builds/56333253

I am not sure what might be the problem on the build 92.6, thanks :-)

**Comment by blink1073** [comment]
@goanpeca, to get PySide to work you use `backend: qt4agg` and `backend.qt4 = pyside`.  See https://github.com/scikit-image/scikit-image/blob/master/tools/travis_script.sh#L107.

**Comment by goanpeca** [comment]
@blink1073 thanks, but why this needs to be done? doesnt spyder take care of this normally?

**Comment by blink1073** [comment]
That's a good question.  Somewhere the `matplotlibrc` is being set up on our behalf.  You could try a `python -c "import matplotlib"` prior to the bootstrap to narrow it down to Spyder itself.

**Comment by ccordoba12** [comment]
@goanpeca, I switched on the .yml option.

About mpl and PySide: we don't set the mpl backend *globally*, just for our Python consoles through our `sitecustomize.py`. I don't know if that answers your question.

How much is left to merge this one?

**Comment by goanpeca** [comment]
I have to see what generates the error with that build (it only happens on that build) and then we should be almost ready to go.

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  - "export DISPLAY=:99.0"
+  - "sh -e /etc/init.d/xvfb start"
+
+install:
+  - ./continuous_integration/travis_install.sh;
+  - export PATH="$HOME/miniconda/bin:$PATH";
+  - source activate test-environment;
+  - QT_API=$USE_QT_API;
+  - echo $QT_API;
+  - python setup.py install;
```
Comment (ccordoba12) [inline .travis.yml]: This is not necessary because you're using bootstrap to start Spyder, so I think it should be removed.

Comment (goanpeca) [inline .travis.yml]: True, maybe we should run both from bootstrap, and then from installed... that way we can check that it is installing correctly, no?

Comment (ccordoba12) [inline .travis.yml]: Mmmhh, but then you should need to add the `--test` flag again to the app cli options.

Comment (goanpeca) [inline .travis.yml]: hmmmmm right...  now, would it make sense to test the install or not ???

Comment (ccordoba12) [inline .travis.yml]: For now we should just use bootstrap. Testing that Spyder installs should be done separately.

**Comment by ccordoba12** [comment]
The PySide error doesn't seems to do with matplotlib (mpl just issues a warning) but with conda package manager.

Let me remove it to see if that fixes the problem.

**Comment by goanpeca** [comment]
ahhh true, I never tried the conda package thinguie with pyside on python 3....

**Comment by ccordoba12** [comment]
Ok, I removed our conda plugin, so please try again to see if everything is passing now :-)

**Comment by goanpeca** [comment]
Ok, but I need to rebase my branch?

**Comment by ccordoba12** [comment]
I don't know how travis works exactly :-) You shouldn't in principle (because there are no conflicts between removing conda and introducing travis.yml) but you probably should so that travis tests Spyder with the new changes

**Comment by goanpeca** [comment]
Ok, I will do this later

**Comment by ccordoba12** [comment]
Could you finish this now? I'd like to merge other PRs and (because I didn't deactivate the yml option) they appear as Failed in the GH interface.

I didn't like to have this message for future references :-)

**Comment by goanpeca** [comment]
There is nothing that can be done with those messages, as they where generated where there was no actual yml to test, if you want to get those messages out, people would need to rebase after this PR is merged, so that their branches include the yml etc....

**Comment by ccordoba12** [comment]
Ok, no biggie then :-) But I really like to merge this one ASAP!

**Comment by blink1073** [comment]
@goanpeca, you'll have to rebase to pick up those changes, because your branch still includes the conda package.

**Comment by goanpeca** [comment]
Ok rebased...

but now I get this error on build [8.2](https://travis-ci.org/spyder-ide/spyder/jobs/56633139)

```bash    
from PySide.QtGui import *                                # analysis:ignore
ImportError: /home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/PySide/QtGui.so: undefined symbol: _ZN8QPainter10drawPixmapERK6QRectFRK7QPixmapS2_

```

I found this when looking at google https://groups.google.com/a/continuum.io/forum/#!topic/anaconda/sUWNlhLykAU

And since anaconda recently updated to 2.2 (minconda maybe did an update?) this has to do something with that.

Comments @ccordoba12 , @blink1073 ?

**Comment by ccordoba12** [comment]
That one is my fault ;-) I updated Qt on Anaconda, so the PySide packages need to be rebuilt against that new Qt version (we missed this fact, so thanks for letting us know about it :-)

I'll try to do that ASAP, but for now please disable the PySide/conda combination (until I let you know that is working again).

**Comment by ccordoba12** [comment]
Even better would be to disable PySide/conda completely because it's not supported by Continuum anymore :-)

**Comment by goanpeca** [comment]
Ok,

In this build https://travis-ci.org/spyder-ide/spyder/jobs/56633143, spyder hsould have toped after 30 seconds...  but it does not... I have no idea why!!!

**Comment by goanpeca** [comment]
And yes, it was the conda package manager... but still puzzled... 

so pyside + 2.7 in non conda install does not shutdown
so pyside + 3.4 in non conda install does shutdown

wtf?!

**Comment by ccordoba12** [comment]
Did you test it locally? Is it working well in that case?

**Comment by goanpeca** [comment]
No I havent, will do now, but it was working 5 days ago!!!!, hmmm I will change the wheelhouse for the pyside install to this one https://github.com/matthew-brett/travis-pyside-builder

I will do this tomorrow, bed time!

**Comment by ccordoba12** [comment]
Yes, don't worry :-)

**Comment by ccordoba12** [comment]
Instead of wheels, why don't you use a ppa?

**Comment by goanpeca** [comment]
Travis uses virtual environments for testing... 

I cannot use `sudo apt-get install python3-pyside` if I want to test inside the virtual envs...

Unless I am missing something

**Comment by blink1073** [comment]
@goanpeca, we use Matthew Brett's wheels in skimage:  https://github.com/scikit-image/scikit-image/blob/master/tools/travis_script.sh#L43

WHEELHOUSE="--no-index --find-links=http://travis-wheels.scikit-image.org/"

**Comment by Nodd** [comment]
@ccordoba12 I thought that PySide can still be installed with conda (but it's not by default), in that case we have to support the combination.
Or has Continuum completely abandoned PySide ?

**Comment by blink1073** [comment]
@Nodd, PySide is available in Conda for Python 2, but not Python 3.

**Comment by goanpeca** [comment]
Why wont continuum make a conda package for pyside and Python3?

**Comment by Nodd** [comment]
Well, threads like this one show that pyside development has gone very slow: http://lists.qt-project.org/pipermail/pyside/2015-February/002251.html

**Comment by goanpeca** [comment]
@nodd Yes I have seen the thread, but if you are going to offer PySide for Python2 why not for Python3 ?

Why not just say that "support of pyside is over" and take all of it out of continuum repos, instead of offering half of it... if there are [wheels that work](https://8167b5c3a2af93a0a9fb-13c6eee0d707a05fa610c311eec04c66.ssl.cf2.rackcdn.com/), why not conda packages?

**Comment by ccordoba12** [comment]
@goanpeca, I can answer that because I'm in charge of Qt/PyQt at Continuum:

1. Anaconda started using PySide (for licensing issues). 
2. The only available Python version at the beginning was Python 2.
3. Anaconda maintainers (not me) noticed that PySide was very unstable (because at that time it was at version 1.1), so they switched to use PyQt4.
4. Then it came Anaconda for Python 3. At this time, PySide was not used anymore in the official installers, so no one felt the need to create a package for this Python version.

The end of the history is this: Continuum is not committed to support PySide anymore because PyQt4 are more stable and better maintained Qt bindings than PySide. I think the PySide package will continue to remain in the repos because of historical reasons and nothing more.

So instead of keep complaining about it, please remove the Conda/PySide combo from the testing matrix. We can still test PySide using Ubuntu, I can't see any really problem with that.

Even if we found problems with conda and PySide, no one at Continuum (meaning me and the Anaconda team) would care about it. They would just say: "please PyQt4 instead".

**Comment by goanpeca** [comment]
@ccordoba12 I do not have a problem in removing it from the matrix.

I use anaconda and I use pyside by default, so it is very annoying that after doing 
```
conda update conda
conda update anaconda
```
Spyder goes @#%#$^$%^#%... so yes I should be able to complain *"So instead of keep complaining about it, "* is unnecessary

It would be better if they (continuum) did not offer pyside support at all instead of a half baked solution.

**Comment by ccordoba12** [comment]
I'll talk to them but I wouldn't hold my breath. If you are using conda, you should be using PyQt4, that's all :-)

Could you comment PySide from .travis.yml so we can merge this one? Then you can add it to in another PR when you make it work again.

**Comment by goanpeca** [comment]
@ccordoba12 give me a sec, I am doing the travis runs on a separate branch just to check that it runs before making the PR....

**Comment by goanpeca** [comment]
Ok, I guess now it should be fine....

[build results](https://travis-ci.org/spyder-ide/spyder/builds/56952270)

**Comment by goanpeca** [comment]
@ccordoba12 since you are the only one doing direct commits to the repo, it would be better if you used `[ci skip]` somewhere in the commit message (not the first line...!) and maybe only leave testing for PRs to avoid making too many build tests? not sure how fair use is defined in travis-ci

http://docs.travis-ci.com/user/how-to-skip-a-build/

**Comment by ccordoba12** [comment]
For that to work I should start to do PRs for everything I do, instead of committing directly to the repo.

How else should we guarantee that Spyder is passing the tests all the time if not? ;-)

File: continuous_integration/travis_install.sh [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  fi
+  pip install -U $TEST_PACKAGES
+}
+
+if [ "$USE_CONDA" = true ] ; then
+  export TEST_PACKAGES="ipython-qtconsole matplotlib pandas pep8 psutil pyflakes pygments pylint sphinx sympy"
+  export SOURCE=`source activate test-environment`
+  install_conda;
+else
+  export TEST_PACKAGES="IPython jedi matplotlib pandas pep8 psutil pyflakes pygments pylint sphinx sympy"
```
Comment (ccordoba12) [inline continuous_integration/travis_install.sh]: After seeing the travis build logs, I think this is causing matplotlib (and all the other packages) to be installed with `pip` instead of using `apt-get`. Two questions:

1. Why not use `apt-get`, at least for mpl? The other packages are just pure Python packages, so there is no problem with using `pip`.
2. If that's not possible, can't use a wheel for mpl?

I say it because the builds are taking a huge amount of time to compile mpl, which is reasonable because mpl is a big package. But we shouldn't build mpl if we can avoid it :-)

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      os: linux
+
+before_install:
+  - sudo apt-get update
+  - "export DISPLAY=:99.0"
+  - "sh -e /etc/init.d/xvfb start"
+
+install:
+  - ./continuous_integration/travis_install.sh;
+  - export PATH="$HOME/miniconda/bin:$PATH";
```
Comment (ccordoba12) [inline .travis.yml]: If you are using an environment, this line is not needed. Why is it here?

File: .travis.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  - source activate test-environment;
+  - QT_API=$USE_QT_API;
+  - echo $QT_API;
```
Comment (ccordoba12) [inline .travis.yml]: This line is unnecessary. `bootstrap.py` reports the Qt toolkit used.

**Comment by ccordoba12** [comment]
I think we should reduce the testing matrix, and just use conda for now. And no, it's not because I'm working at Continuum :-), it's because conda packages are (almost always) up to date and quite reliable.

I also say it because the wheels in http://travis-wheels.scikit-image.org/ seem outdated (there is no mpl 1.4.3 for example).

At the end, the tests using conda are already passing, and do so only taking 1 min or so. So why not start with that and leave the other options for later?

File: continuous_integration/travis_install.sh (hunk 5 above)
Comment (goanpeca) [inline continuous_integration/travis_install.sh]: 1. Travis uses environments (with virtual env), so using apt-get will install packages for the global python which is 2.7 and that makes no sense for python 3.3, and 3.4. Travis VMs are all the same, what changes is the virtual envs....
2. Yes, we can use wheels (if they are available...) for all the other packages that take too long too compile. Otherwise we would need to make our own wheels for the different packages we are testing.

File: .travis.yml (hunk 6 above)
Comment (goanpeca) [inline .travis.yml]: Cause I cannot activate an environment from within a bash file, so this line when using conda will activate the environment, and when not using conda wont do anything... that is why

File: .travis.yml (hunk 7 above)
Comment (goanpeca) [inline .travis.yml]: Yes, I left it there before even bootstrap was running... will remove

File: .travis.yml (hunk 6 above)
Comment (ccordoba12) [inline .travis.yml]: Not exactly, but I see its purpose now. This line is needed to add `miniconda/bin` to PATH so that `source activate` works.

**Comment by goanpeca** [comment]
ok

File: continuous_integration/travis_install.sh (hunk 5 above)
Comment (ccordoba12) [inline continuous_integration/travis_install.sh]: Then (as I said), please use a wheel for mpl. The other packages are fine.

File: .travis.yml (hunk 7 above)
Comment (ccordoba12) [inline .travis.yml]: ok

File: .travis.yml (hunk 6 above)
Comment (goanpeca) [inline .travis.yml]: Ahh I thought you meant the one after... and yes... it is needed for that... becuase for some reason when doing the same inside the bash script it was not working.. so I took it out... and it worked...

**Comment by goanpeca** [comment]
So for the moment only conda... when we have a set of wheels ready and updated we can activate the rest

File: continuous_integration/travis_install.sh (hunk 5 above)
Comment (goanpeca) [inline continuous_integration/travis_install.sh]: For the moment only conda then.... I will add the wheels later on...

**Comment by ccordoba12** [comment]
Yes, that's ok.

One last question: what do you recommend for *my* particular workflow? Should I start making PRs instead of pushing directly to master?

**Comment by blink1073** [comment]
I agree with using all conda all the time, and letting you do whatever is easiest for you @ccordoba12.  For instance, minrk pushes directly to master on IPython quite often.

**Comment by goanpeca** [comment]
Lets just keep doing things as usual and see how it goes :smile:.

Is this ready for merging, or is there something I need to fix before?

I am starting with the tests now, specifically the source code editor.

Cheers

**Comment by ccordoba12** [comment]
Ok, merging then! Thanks a lot for your work on this @goanpeca, this is immensely useful!

About the other tests, please start simple. Let's start with the Editor in one PR, and then the other plugins in further PRs.

**Comment by SylvainCorlay** [comment]
Sorry I am late in the battle. I have the same issue when testing qtawesome with PySide: undefined symbol when trying to import it. Although I still think that supporting PySide is valuable, especially license-wise. Should I drop it from the conda tests?

**Comment by goanpeca** [comment]
@SylvainCorlay, no battle here ;-), sure disabled (comment the lines...) I will add pyside from normal installs (using wheels) later

Cheers

**Comment by SylvainCorlay** [comment]
@goanpeca it is just an idiom. / I hope anaconda will support pyside again in the future.

**Comment by ccordoba12** [comment]
The official answer is that Continuum doesn't support PySide anymore.

However, I'll try to update it on my free time.
